### PR TITLE
PUBDEV-3971: Add cross-validation to Stacked Ensemble

### PR DESCRIPTION
**Pull request by ledell**
~Work in progress.~  Ready for review.

### CODE REVIEW

File: h2o-algos/src/main/java/hex/StackedEnsembleModel.java [hunk 1]
```diff
@@ -29,9 +29,13 @@
...
   private NonBlockingHashSet<String> names = null;  // keep columns as a set for easier comparison
-  public int nfolds = -1; //From 1st base model
-  public Parameters.FoldAssignmentScheme fold_assignment; //From 1st base model
-  public String fold_column; //From 1st base model
+  //public int nfolds = -1; //From 1st base model
+
+  // Get from 1st base model (should be identical across base models)
+  public int basemodel_nfolds = -1;  //From 1st base model
+  public Parameters.FoldAssignmentScheme basemodel_fold_assignment;  //From 1st base model
+  public String basemodel_fold_column;  //From 1st base model
```
Comment (navdeep-G) [inline h2o-algos/src/main/java/hex/StackedEnsembleModel.java]: Like the name changes

File: h2o-r/tests/testdir_algos/stackedensemble/runit_stackedensemble_nfolds.R [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                                y = y,
+                                training_frame = train,
+                                base_models = list(my_gbm, my_rf))
+  # Check that metalearner_nfolds is correctly stored in model output
+  expect_equal(stack0@parameters$metalearner_nfolds, NULL)
+  expect_equal(stack0@allparameters$metalearner_nfolds, 0)
+  # Check that the metalearner was cross-validated with the correct number of folds
+  meta0 <- h2o.getModel(stack0@model$metalearner$name)
+  expect_equal(meta0@parameters$nfolds, NULL)
+  expect_equal(meta0@allparameters$nfolds, 0)
```
Comment (navdeep-G) [inline h2o-r/tests/testdir_algos/stackedensemble/runit_stackedensemble_nfolds.R]: Why the diff between `parameters` and `allparameters` again?

Comment (ledell) [reply h2o-r/tests/testdir_algos/stackedensemble/runit_stackedensemble_nfolds.R]: `parameters` stores only the named parameters set explicitly by the user, where as `allparameters` stores all the parameters and their values (default values with the exception of those you specified explicitly).

Comment (ledell) [reply h2o-r/tests/testdir_algos/stackedensemble/runit_stackedensemble_nfolds.R]: In this particular instance, it's checking that the metalearner's `nfolds` was not set (NULL) and that the value of the parameter internally is still 0.

Comment (navdeep-G) [reply h2o-r/tests/testdir_algos/stackedensemble/runit_stackedensemble_nfolds.R]: Ah, okay. Makes sense.

File: h2o-algos/src/main/java/hex/StackedEnsembleModel.java [hunk 3]
```diff
@@ -178,14 +181,16 @@ public ModelMetrics doScoreMetricsOneFrame(Frame frame, Job job) {
...
+    // To get ensemble training metrics, the training frame needs to be re-scored since
+    // training metrics from metalearner are not equal to ensemble training metrics
     this._output._training_metrics = doScoreMetricsOneFrame(this._parms.train(), job);
-    if (null != this._parms.valid()) {
-      this._output._validation_metrics = doScoreMetricsOneFrame(this._parms.valid(), job);
-    }
+    // Validation and cross-validation metrics can be copied from metalearner (may be null)
+    this._output._validation_metrics = this._output._metalearner._output._validation_metrics;
+    this._output._cross_validation_metrics = this._output._metalearner._output._cross_validation_metrics;
   }
```
Comment (navdeep-G) [inline h2o-algos/src/main/java/hex/StackedEnsembleModel.java]: The copy vs no copy should be documented

File: h2o-algos/src/main/java/hex/schemas/StackedEnsembleV99.java [hunk 4]
```diff
@@ -5,32 +5,50 @@
...
     public KeyV3.ModelKeyV3 base_models[];
 
     @API(help = "Keep level one frame used for metalearner training.")
     public boolean keep_levelone_frame;
+
+    @API(level = API.Level.critical, direction = API.Direction.INOUT, gridable = true,
+            help = "Number of folds for K-fold cross-validation of the metalearner algorithm (0 to disable or >= 2).")
+    public int metalearner_nfolds;
+
+    // TO DO: Add metalearner_fold_column
```
Comment (navdeep-G) [inline h2o-algos/src/main/java/hex/schemas/StackedEnsembleV99.java]: Might be good to tag the JIRA ticket here.

File: h2o-py/h2o/estimators/estimator_base.py [hunk 5]
```diff
@@ -241,7 +241,8 @@ def _resolve_model(self, model_id, model_json):
                         model_json["output"][metric] = \
                             metrics_class(model_json["output"][metric], metric, model_json["algo"])
 
-            if m._is_xvalidated:
+            #if m._is_xvalidated:
+            if m._is_xvalidated and model_json["output"]["cross_validation_models"] is not None:
```
Comment (navdeep-G) [inline h2o-py/h2o/estimators/estimator_base.py]: Nice catch

**Comment by navdeep-G** [comment]
Just some minor comments. Rest looks good. Lets wait for the CI to complete.

File: h2o-algos/src/main/java/hex/StackedEnsembleModel.java (hunk 3 above)
Comment (ledell) [reply h2o-algos/src/main/java/hex/StackedEnsembleModel.java]: Documented how/where?

File: h2o-algos/src/main/java/hex/schemas/StackedEnsembleV99.java (hunk 4 above)
Comment (ledell) [reply h2o-algos/src/main/java/hex/schemas/StackedEnsembleV99.java]: Ok sure, will do.

File: h2o-algos/src/main/java/hex/StackedEnsembleModel.java (hunk 3 above)
Comment (navdeep-G) [reply h2o-algos/src/main/java/hex/StackedEnsembleModel.java]: Just in the code for devs.

**Comment by ledell** [comment]
@h2o-ops please test!

**Pull request merged**